src/form/iv.jl

```diff
@@ -566,3 +378,48 @@ function _objective_min_fuel_and_flow_cost_polynomial_linquad(pm::AbstractIVRMod
         for (n, nw_ref) in nws(pm))
     )
 end
+
+
+"adds pg_cost variables and constraints"
+function objective_variable_pg_cost(pm::AbstractIVRModel)
+    for (n, nw_ref) in nws(pm)
+        gen_lines = calc_cost_pwl_lines(nw_ref[:gen])
+
+        #to avoid function calls inside of @NLconstraint
+        pg_cost = var(pm, n)[:pg_cost] = JuMP.@variable(pm.model,
+            [i in ids(pm, n, :gen)], base_name="$(n)_pg_cost",
+        )
+        nc = length(conductor_ids(pm, n))
+
+        # gen pwl cost
+        for (i, gen) in nw_ref[:gen]
+            pg = [var(pm, n, c, :pg, i) for c in conductor_ids(pm, n)]
+            for line in gen_lines[i]
+                JuMP.@NLconstraint(pm.model, pg_cost[i] >= line.slope*sum(pg[c] for c in 1:nc) + line.intercept)
+            end
+        end
+    end
+end
+
+
+"adds p_dc_cost variables and constraints"
+function objective_variable_dc_cost(pm::AbstractIVRModel)
+    for (n, nw_ref) in nws(pm)
+        dcline_lines = calc_cost_pwl_lines(nw_ref[:dcline])
+
+        dc_p_cost = var(pm, n)[:p_dc_cost] = JuMP.@variable(pm.model,
+            [i in ids(pm, n, :dcline)], base_name="$(n)_dc_p_cost",
+        )
+        #to avoid function calls inside of @NLconstraint:
+        nc = length(conductor_ids(pm, n))
+        # dcline pwl cost
+        for (i, dcline) in nw_ref[:dcline]
+            arc = (i, dcline["f_bus"], dcline["t_bus"])
+            for line in dcline_lines[i]
+                #to avoid function calls inside of @NLconstraint:
+                p_dc = [var(pm, n, c, :p_dc, arc) for c in conductor_ids(pm, n)]
+                JuMP.@NLconstraint(pm.model, dc_p_cost[i] >= line.slope*sum(p_dc[c] for c in 1:nc)  + line.intercept)
```

Comment:
Same note from other objective.

Reply:
:p_dc contains @NLexpressions, which can be nested into @NLconstraints only.